src/node_report_module.cc

```diff
@@ -0,0 +1,539 @@
+      fprintf(stdout,
+              "node-report: SignalDumpInterruptCallback handling signal\n");
+    }
+    if (nodereport_events & kReportOnSignal) {
+      if (nodereport_verbose) {
+        fprintf(stdout,
+                "node-report: SignalDumpInterruptCallback triggering report\n");
+      }
+      TriggerNodeReport(isolate,
+                        Environment::GetCurrent(isolate),
+                        kSignal_JS,
+                        node::signo_string(report_signal.load()),
+                        __func__,
+                        nullptr,
+                        MaybeLocal<Value>());
+    }
+    report_signal = 0;
+  }
+}
+
+static void SignalDumpAsyncCallback(uv_async_t* handle) {
+  Isolate* isolate = Isolate::GetCurrent();
+  Environment* env = Environment::GetCurrent(isolate);
+  CHECK_NE(env, nullptr);
+  if (report_signal.load() != 0) {
+    if (nodereport_verbose) {
+      fprintf(stdout,
+              "node-report: SignalDumpAsyncCallback handling signal\n");
+    }
+    if (nodereport_events & kReportOnSignal) {
+      if (nodereport_verbose) {
+        fprintf(stdout,
+                "node-report: SignalDumpAsyncCallback triggering NodeReport\n");
+      }
+      TriggerNodeReport(isolate, env,
+                        kSignal_UV,
+                        node::signo_string(report_signal.load()),
+                        __func__,
+                        nullptr,
+                        MaybeLocal<Value>());
+    }
+    report_signal = 0;
+  }
+}
+
+/*******************************************************************************
+ * Utility functions for signal handling support (platforms except Windows)
+ *  - RegisterSignalHandler() - register a raw OS signal handler
+ *  - SignalDump() - implementation of raw OS signal handler
+ *  - StartWatchdogThread() - create a watchdog thread
+ *  - ReportSignalThreadMain() - implementation of watchdog thread
+ *  - SetupSignalHandler() - initialisation of signal handlers and threads
+ ******************************************************************************/
+// Utility function to register an OS signal handler
+/*
+static void RegisterSignalHandler(int signo, void (*handler)(int),
+                                  struct sigaction* saved_sa) {
+  struct sigaction sa;
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = handler;
+  sigfillset(&sa.sa_mask);  // mask all signals while in the handler
+  sigaction(signo, &sa, saved_sa);
+}
+*/
+
+// Utility function to restore an OS signal handler to its previous setting
+static void RestoreSignalHandler(int signo, struct sigaction* saved_sa) {
+  sigaction(signo, saved_sa, nullptr);
+}
+
+// Raw signal handler for triggering a report - runs on an arbitrary thread
+static void SignalDump(int signo) {
+  // Check atomic for report already pending, storing the signal number
+  if (report_signal.exchange(signo) == 0) {
+    uv_sem_post(&report_semaphore);  // Hand-off to watchdog thread
+  }
+}
```

Comment:
Yes, the idea is to trigger a report at arbitrary program points through a signal, and let the program continue (non-invasive).

doing it through JS:
```js
process.on(REPORTSIGNAL, (sig) => {
  util.report.getReport();
});
```

is what you mean? if so, that looks pretty straightforward, and easy. I am not sure if Windows (where sending signals are not natively supported) is the reason for doing this in C++.

Let me think over this.

Reply:
for unknown reasons, `process.on()` does not have any effect if when placed in `util.js` - its probably it is too early when the event handling is not started on process object or something like that?

Reply:
Sorry I didn’t reply sooner – yes, that’s what I had in mind.

If you’re adding the listener during bootstrap, then yes, that is most likely what happens… are you inspecting the command line arguments for this, and deciding based on that? I guess we could amend our signal handling code to allow for handlers added before its own initialization?

Reply:
some findings in the interim:

[setupSignalHandlers](https://github.com/nodejs/node/blob/0bdb95f4cffa7d672c23c60cdb6b779451d2ea33/lib/internal/process/main_thread_only.js#L120) is  what empowers signals to be treated at libuv level. If we add a listener on a signal before this method was called, it just follows normal event semantics only, and respond only to 'emit' calls, not real signals.

so the reason `process.on()` does not have any effect is understood; its placement is not proper.

I guess its ideal placement is in `lib/internal/bootstrap/node.js` . I tried some positions in `startup()` which I thought was making sense, but looks like it is still ignored.

 - hint 1: process.on('SIGUSR2') in REPL mode works fine and provides callback.
 - hint 2: looks like it may be related to preloadModules, but don't know how.

Reply:
It’s happening relatively early in the overall startup phase, but core modules can still require `util` before this happens.

Maybe we could refactor `setupSignalHandlers()` to account for existing handlers? That seems like a somewhat robust solution that could work.

Reply:
yes, this is done now by accommodating existing handlers in `setupSignalHandlers`. I spent good amount of time debugging why `process.on` does not have any effect at places where I think it should work (for example in `lib/internal/bootstrap/node.js`  post `setupSignalHandlers` ) and finally figured out my tiny test case was the problem - there was nothing else left in the program other than the handler and the code was exiting. Ref: 
https://github.com/nodejs/node/issues/19586

Reply:
this is all resolved now; #24651 covers the support for this.